functions/src/resolve-market.ts

```diff
@@ -31,140 +35,207 @@ export const resolveMarket = functions
-          return { status: 'error', message: 'Invalid outcome' }
-      } else if (outcomeType === 'FREE_RESPONSE') {
-        if (
-          isNaN(+outcome) &&
-          !(outcome === 'MKT' && resolutions) &&
-          outcome !== 'CANCEL'
+      if (contract.creatorId !== userId)
+        return { status: 'error', message: 'User not creator of contract' }
+
+      return privateResolveMarket(contract, data, contractDoc)
+    }
+  )
+
+export const autoResolveMarkets = functions.pubsub
+  .schedule('every 1 hours')
+  .onRun(async () => {
+    const contracts = await getValues<Contract>(
+      firestore.collection('contracts')
+    )
+
+    const contractsToResolve = contracts
+      .filter((c) => !c.isResolved)
+      .filter((c) => c.autoResolutionTime < Date.now())
+
+    await batchedWaitAll(
+      contractsToResolve.map((contract) => async () => {
+        const result = await autoResolve(contract)
+
+        console.log(
+          'resolved',
+          contract.slug,
+          contract.autoResolution,
+          'result:',
+          result
         )
-          return { status: 'error', message: 'Invalid outcome' }
-      } else if (outcomeType === 'NUMERIC') {
-        if (isNaN(+outcome) && outcome !== 'CANCEL')
-          return { status: 'error', message: 'Invalid outcome' }
-      } else {
-        return { status: 'error', message: 'Invalid contract outcomeType' }
-      }
+      })
+    )
+  })
+
+const autoResolve = async (contract: Contract) => {
+  const data = {
+    outcome: contract.autoResolution,
+    value: undefined, // numeric
+    probabilityInt:
+      contract.outcomeType == 'BINARY'
+        ? getProbability(contract) * 100
+        : undefined,
+    resolutions:
+      contract.outcomeType == 'FREE_RESPONSE'
+        ? getFreeResponseResolutions(contract)
+        : undefined,
+  }
+  const contractDoc = firestore.doc(`contracts/${contract.id}`)
+  return await privateResolveMarket(contract, data, contractDoc, true)
+}
 
-      if (value !== undefined && !isFinite(value))
-        return { status: 'error', message: 'Invalid value' }
+const getFreeResponseResolutions = (contract: Contract & FreeResponse) => {
+  const answersWithProbs = getAnswersWithProbs(contract)
+  const totalProb = sum(Object.values(answersWithProbs))
+  return mapValues(answersWithProbs, (prob) => (100 * prob) / totalProb)
+}
 
+const getAnswersWithProbs = (contract: Contract & FreeResponse) => {
+  const answers: { [id: string]: number } = {}
+  for (const answer of contract.answers) {
+    answers[answer.id] = getDpmOutcomeProbability(
+      contract.totalShares,
+      answer.id
+    )
+  }
+  return answers
+}
+
+const privateResolveMarket = async (
+  contract: Contract,
+  data: {
+    outcome: resolution
+    value?: number
+    probabilityInt?: number
+    resolutions?: { [outcome: string]: number }
+  },
+  contractDoc: admin.firestore.DocumentReference<admin.firestore.DocumentData>,
+  autoResolution = false
+) => {
```

Comment:
This whole file kind of seems like a mess to me, but I am not going to say you have to fix it to merge this. It would just be nice if it were nicer.